File: python/2d-dynamic-programming/target-sum.py
class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        cache = {}

        def count(current, i):
            if i == len(nums):
                if current == target:
                    return 1
                else:
                    return 0

            if (current, i) in cache:
                return cache[(current, i)]

            cache[(current, i)] = count(current + nums[i], i + 1) + count(current - nums[i], i + 1)

            return cache[(current, i)]

        return count(0, 0)

File: c++/2d-dynamic-programming/target-sum.cc
class Solution {
public:
    int findTargetSumWays(vector<int>& nums, int target) {
        std::unordered_map<int, std::unordered_map<int, int>> cache;
        return sum(0, target, nums, cache);
    }

private:
    int sum(int i, int target, std::vector<int>& nums, std::unordered_map<int, std::unordered_map<int, int>>& cache) {
        if (i == nums.size() && target == 0) {
            return 1;
        }

        if (i == nums.size()) {
            return 0;
        }

        if (cache.contains(i) && cache[i].contains(target)) {
            return cache[i][target];
        }

        cache[i][target] = sum(i + 1, target - nums[i], nums, cache) + sum(i + 1, target + nums[i], nums, cache);

        return cache[i][target];
    }
};
